[HttpKernel] LoggerDataCollector: splitting logs on different sub-requests

vtsykun opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | -
| Fixed tickets | #23620 
| License       | MIT
| Doc PR        | -

This PR fixed performance problem in dev mode and prevent logs serialize several times for each sub-request.

STR:
1) Install any symfony application v2.8-3.4
2) Create a simple page with 20-50 sub-requests and 5k-10k logs
3) Open page in dev mode

**Actual**
- debug toolbar not open (404 not found)
- sets of logs (in the logger debug panel) for different requests are same
- log processing takes about 20-30s
- gc also run about 50% of execution time


**Expected**
- Debug toolbar should be open
- As developer, I want to see in logger panel only those logs that are relevant to the given sub-request

### Profile and performance

Tested on page with 7k logs and 70 sub-request.

Comparison:
- v3.3.5 - this patch: https://blackfire.io/profiles/compare/b1d410b3-c4a7-4778-9b6d-514f72193e28/graph
- v3.3.5 - patch #23644 https://blackfire.io/profiles/compare/d0715b04-7834-4981-8da2-9f1dcb2ef90c/graph

nicolas-grekas commented:
Should the profile for the master request get all logs? At least yes when `framework.profiler.only_master_requests` is set.
Thinking about the toolbar/panel, having the total count for logs of all sub requests looks desirable to me also. On the implementation side, a simple implementation would require cloning values from collector twice. A better implem would looks for profiles of all subrequests.

javiereguiluz commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -45,6 +66,15 @@ public function __invoke(array $record)
+        if (func_num_args() === 1) {
+            $requestHash = func_get_arg(0);
+            if (isset($this->recordsRequest[$requestHash]['records'])
+                && $requestHash !== $this->masterRequestHash
```
We usually don't split these long `if()` conditions, so this should be one line.

vtsykun replied:
Done

javiereguiluz commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -45,6 +66,15 @@ public function __invoke(array $record)
      */
     public function getLogs()
     {
+        if (func_num_args() === 1) {
```
In Symfony code we use [Yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions), so this should be `1 === func_num_args ()` instead.

vtsykun replied:
Fixed

javiereguiluz commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -53,6 +83,27 @@ public function getLogs()
+        if (func_num_args() === 1) {
+            $requestHash = func_get_arg(0);
+            if (isset($this->recordsRequest[$requestHash]['errorCount'])
+                && $requestHash !== $this->masterRequestHash
```
same with this `if()` condition

vtsykun commented:
I guess that profile should for the master request get all logs

vtsykun committed:
[HttpKernel] LoggerDataCollector: splitting logs on different sub-requests

nicolas-grekas commented:
> I guess that profile should for the master request get all logs

@vtsykun same to me. OK to finish this PR then?

nicolas-grekas commented:
ping @vtsykun

vtsykun commented:
Hi @nicolas-grekas, this done, please review 
Yes, I updated PR and squashed commit on Jul 29. So in this implementation I show all logs for master request.

ro0NL commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -45,14 +63,50 @@ public function __invoke(array $record)
+    }
+
+    /**
+     * @param GetResponseEvent $event
```
can be removed i guess

ro0NL commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -45,14 +63,50 @@ public function __invoke(array $record)
+            return;
+        }
+
+        $this->masterRequestHash = spl_object_hash($event->getRequest());
```
might store the current request hash as well (to use if not passed on the calling side)

ro0NL replied:
is passing it on the calling side really needed?

vtsykun replied:
I thinking the better move it on `LoggerDataCollector` side.

ro0NL commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -12,29 +12,47 @@
 class DebugProcessor implements DebugLoggerInterface
 {
-    private $records = array();
-    private $errorCount = 0;
```
what about array for both? in favor of `array_column`

ro0NL reviewed (commented):
nice :+1:

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/LoggerDataCollector.php:
```diff
@@ -25,22 +26,30 @@ class LoggerDataCollector extends DataCollector implements LateDataCollectorInte
+        if (null !== $this->requestStack && $this->requestStack->getMasterRequest() !== $request) {
+            $this->currentRequestHash = spl_object_hash($request);
+        } else {
+            //gets all logs for the master request
```
space missing, but is obvious to me

ro0NL commented on src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php:
```diff
@@ -45,14 +62,36 @@ public function __invoke(array $record)
+        }
+
+        if (1 === func_num_args()) {
+            $requestHash = func_get_arg(0);
```
collector might pass `null` here

vtsykun replied:
Fixed.

ro0NL reviewed (commented):
almost

fabpot commented on src/Symfony/Bridge/Monolog/Logger.php:
```diff
@@ -27,7 +27,7 @@ class Logger extends BaseLogger implements DebugLoggerInterface
     {
         if ($logger = $this->getDebugLogger()) {
-            return $logger->getLogs();
+            return call_user_func_array(array($logger, 'getLogs'), func_get_args());
```
Could be replace with: `$logger->getLogs(...func_get_args());` which is easier to read and faster.

ro0NL replied:
not php 5.5 compat

vtsykun replied:
Yes, it not supported on php < 5.6 http://sandbox.onlinephpfunctions.com/code/3ba4ee65a58af1267040434818d752f26bc885a1

fabpot commented on src/Symfony/Bridge/Monolog/Logger.php:
```diff
@@ -39,7 +39,7 @@ public function getLogs()
     {
         if ($logger = $this->getDebugLogger()) {
-            return $logger->countErrors();
+            return call_user_func_array(array($logger, 'countErrors'), func_get_args());
```
same here

ro0NL commented:
From #24275; do events have this same issue?

nicolas-grekas commented:
@vtsykun thank you for this PR, and sorry I didn't manage to review it earlier. I just rebased it on top of the latest version of branch 3.4, with a few minor changes. The code looks to good to me. It sill misses a few unit tests. Would you be up for adding them? If yes, don't forget to fetch+rebase --hard on your remote patch-splitting-logs branch. If not, I'm sure someone else can take over. Please advise.

nicolas-grekas reviewed (approved):
(I just pushed some unit tests)

fabpot reviewed (approved):
Looks like something for 4.1/master intead

nicolas-grekas commented:
> Looks like something for 4.1/master instead

It could go on 4.1 for sure. The only thing that made me keep it for 3.4 is that the reported behavior affects 3.4 (sub-requests heavy pages suffer from a significant penalty because of log duplication.)

nicolas-grekas commented:
Rebased on master, ready for 4.1.

nicolas-grekas commented:
Thank you @vtsykun.
